python/lsst/daf/butler/formatters/parquet.py

```diff
@@ -883,3 +929,145 @@ def _append_numpy_string_metadata(metadata: Dict[bytes, str], name: str, dtype:
         metadata[f"lsst::arrow::len::{name}".encode("UTF-8")] = str(dtype.itemsize // 4)
     elif dtype.type is np.bytes_:
         metadata[f"lsst::arrow::len::{name}".encode("UTF-8")] = str(dtype.itemsize)
+
+
+def _append_numpy_multidim_metadata(metadata: Dict[bytes, str], name: str, dtype: np.dtype) -> None:
+    """Append numpy multi-dimensional shapes to arrow metadata.
+
+    All column types are handled, but the metadata is only modified for
+    multi-dimensional columns.
+
+    Parameters
+    ----------
+    metadata : `dict` [`bytes`, `str`]
+        Metadata dictionary; modified in place.
+    name : `str`
+        Column name.
+    dtype : `np.dtype`
+        Numpy dtype.
+    """
+    if len(dtype.shape) > 1:
+        metadata[f"lsst::arrow::shape::{name}".encode("UTF-8")] = str(dtype.shape)
+
+
+def _multidim_shape_from_metadata(metadata: Dict[bytes, bytes], list_size: int, name: str) -> tuple[int, ...]:
+    """Retrieve the shape from the metadata, if available.
+
+    Parameters
+    ----------
+    metadata : `dict` [`bytes`, `bytes`]
+        Metadata dictionary.
+    list_size : `int`
+        Size of the list datatype.
+    name : `str`
+        Column name.
+
+    Returns
+    -------
+    shape : `tuple` [`int`]
+        Shape associated with the column.
+
+    Raises
+    ------
+    RuntimeError
+        Raised if metadata is found but has incorrect format.
+    """
+    md_name = f"lsst::arrow::shape::{name}"
+    if (encoded := md_name.encode("UTF-8")) in metadata:
+        groups = re.search(r"\((.*)\)", metadata[encoded].decode("UTF-8"))
+        if groups is None:
+            raise RuntimeError("Illegal value found in metadata.")
+        shape = tuple(int(x) for x in groups[1].split(",") if x != "")
+    else:
+        shape = (list_size,)
+
+    return shape
+
+
+def _schema_to_dtype_list(schema: pa.Schema) -> list[tuple[str, tuple[Any] | str]]:
+    """Convert a pyarrow schema to a numpy dtype.
+
+    Parameters
+    ----------
+    schema : `pyarrow.Schema`
+
+    Returns
+    -------
+    dtype_list: `list` [`tuple`]
+        A list with name, type pairs.
+    """
+    metadata = schema.metadata if schema.metadata is not None else {}
+
+    dtype = []
+    for name in schema.names:
+        t = schema.field(name).type
+        if isinstance(t, pa.FixedSizeListType):
+            shape = _multidim_shape_from_metadata(metadata, t.list_size, name)
+            dtype.append((name, (t.value_type.to_pandas_dtype(), shape)))
+        elif t not in (pa.string(), pa.binary()):
+            dtype.append((name, t.to_pandas_dtype()))
+        else:
+            dtype.append((name, _arrow_string_to_numpy_dtype(schema, name)))
+
+    return dtype
+
+
+def _numpy_dtype_to_arrow_types(dtype: np.dtype) -> list[Any]:
+    """Convert a numpy dtype to a list of arrow types.
+
+    Parameters
+    ----------
+    dtype : `numpy.dtype`
+        Numpy dtype to convert.
+
+    Returns
+    -------
+    type_list : `list` [`object`]
```

Comment:
I think Sphinx really wants a description after this, even if it's redundant with the initial line of the docstring.